File: test_cases/cpp_turnstile/turnstile_test.cc
#include <stdio.h>
#include <strings.h>
#include <string.h>
#include <stdlib.h>

#include "TurnstileActions.h"
#include "TwoCoinTurnstile.h"

static char output[80];
static int position;
static int error_count = 0;

class MyTwoCoinTurnstile : public TwoCoinTurnstile {
public:		
	virtual void lock() {
		output[position++] = 'L';
	}

	void unlock() {
		output[position++] = 'U';
	}

	void thankyou() {
		output[position++] = 'T';
	}

	void alarmOn() {
		output[position++] = 'A';
	}

	void alarmOff() {
		output[position++] = 'O';
	}

	void unexpected_transition (const char* state, const char* event) {
		char error[80];
		bzero(error, 80);
		sprintf(error, "X(%s,%s)", state, event);
		strcpy(output+position, error);
		position += strlen(error);
	}
};

static MyTwoCoinTurnstile* fsm;

static void setup() {
	bzero(output, 80);
	position = 0;
		
	fsm = new MyTwoCoinTurnstile();
}

void check(const char* function, const char* expected) {
	int result = strcmp(output, expected);
	if (result) {
		printf("\n%s failed.  expected: %s, but was: %s", function, expected, output);
		error_count++;
	}
	else
		printf(".");	
}

static void testNormalBehavior() {
	setup();
	fsm->Coin();
	fsm->Coin();
	fsm->Pass();
	check("testNormalBehavior", "UL");
}

static void testAlarm() {
	setup();
	fsm->Pass();
	check("testAlarm", "A");
}

static void testThankyou() {
	setup();
	fsm->Coin();
	fsm->Coin();
	fsm->Coin();
	check("testThankyou", "UT");
}

static void testNormalManyThanksAndAlarm() {
	setup();
	fsm->Coin();
	fsm->Coin();
	fsm->Pass();
	fsm->Coin();
	fsm->Coin();
	fsm->Coin();
	fsm->Pass();
	fsm->Pass();
	check("testNormalManyThanksAndAlarm", "ULUTLA");
}


static void testUndefined() {
	setup();
	fsm->Pass();
	fsm->Pass();
	check("testUndefined", "AX(Alarming,Pass)");
}

static void testReset() {
	setup();
	fsm->Pass();
	fsm->Reset();
	check("testReset", "AOL");
}

int main(int ac, char** av) {
	printf("Turnstile test begins\n");
	testNormalBehavior();
	testAlarm();
	testThankyou();
	testNormalManyThanksAndAlarm();
	testUndefined();
	testReset();
	if (error_count)
		printf("\nTests Complete with %d errors.\n", error_count);
	else
		printf("\nOK\n");
}

File: test_cases/go_turnstile/turnstile_test.go
package twocointurnstile

import (
	"fmt"
	"testing"
)

var _ TurnstileActions = &MyTwoCoinTurnstile{}

// MyTwoCoinTurnstile implements the TurnstileActions interface.
type MyTwoCoinTurnstile struct {
	output   string
	position int
	t        *testing.T
}

func (m *MyTwoCoinTurnstile) Lock() {
	m.output += "L"
	m.position++
}

func (m *MyTwoCoinTurnstile) Unlock() {
	m.output += "U"
	m.position++
}

func (m *MyTwoCoinTurnstile) Thankyou() {
	m.output += "T"
	m.position++
}

func (m *MyTwoCoinTurnstile) AlarmOn() {
	m.output += "A"
	m.position++
}

func (m *MyTwoCoinTurnstile) AlarmOff() {
	m.output += "O"
	m.position++
}

func (m *MyTwoCoinTurnstile) UnexpectedTransition(state, event string) {
	e := fmt.Sprintf("X(%v,%v)", state, event)
	m.output += e
	m.position += len(e)
}

func (m *MyTwoCoinTurnstile) check(function, want string) {
	m.t.Helper()

	if m.output != want {
		m.t.Errorf("%s failed: got %v, want %v", function, m.output, want)
	}
}

func setup(t *testing.T) (*TwoCoinTurnstile, *MyTwoCoinTurnstile) {
	m := &MyTwoCoinTurnstile{t: t}
	fsm := New(m)
	return fsm, m
}

func TestNormalBehavior(t *testing.T) {
	fsm, m := setup(t)
	fsm.Coin()
	fsm.Coin()
	fsm.Pass()
	m.check("testNormalBehavior", "UL")
}

func TestAlarm(t *testing.T) {
	fsm, m := setup(t)
	fsm.Pass()
	m.check("testAlarm", "A")
}

func TestThankyou(t *testing.T) {
	fsm, m := setup(t)
	fsm.Coin()
	fsm.Coin()
	fsm.Coin()
	m.check("testThankyou", "UT")
}

func TestNormalManyThanksAndAlarm(t *testing.T) {
	fsm, m := setup(t)
	fsm.Coin()
	fsm.Coin()
	fsm.Pass()
	fsm.Coin()
	fsm.Coin()
	fsm.Coin()
	fsm.Pass()
	fsm.Pass()
	m.check("testNormalManyThanksAndAlarm", "ULUTLA")
}

func TestUndefined(t *testing.T) {
	fsm, m := setup(t)
	fsm.Pass()
	fsm.Pass()
	m.check("testUndefined", "AX(Alarming,Pass)")
}

func TestReset(t *testing.T) {
	fsm, m := setup(t)
	fsm.Pass()
	fsm.Reset()
	m.check("testReset", "AOL")
}
